### PR TITLE
[PretrainedConfig] Fix save pretrained config for edge case 

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

There is an edge case for which the "diff" save method for `PretrainedConfig` fails. We decided a while ago in this PR: https://github.com/huggingface/transformers/pull/3797 that we wanted to have more readable configs and thus tweaked the `save_pretrained()` method so that only parameters that are different to the default **PretrainedConfig** class are serialized. 

There was an edge case we did not consider:

If a parameter, like `add_cross_attention` defaults to `True` in `ProphetNetConfig`, but is by default `False` in `PretrainedConfig` a problem can arise when a user wants to save `add_cross_attention=False` in his `ProphetNetConfig`. Because `add_cross_attention=False` corresponds to the `PretrainedConfig` default case, this parameter will not be serialized and thus when reloading the config, the parameter defaults to `ProphetNetConfig` which is `True` and which is then an error.

This PR fixes this behavior by simply making sure that a parameter is only **not** saved if it is equal to both `PretrainedConfig` and `ProphetNetConfig`. 

This feature requires configs to be instantiated without providing any parameters. This is currently not possible for the `EncoderDecoderModelConfig` and `RagConfig` because those configs are composed of multiple sub-configs which have to be provided. => A new class attribute `is_composition` is added to correctly handle these classes.

Two tests are added.

Also cc @stas00 for FSTM config.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sgugger** [comment]
Any reason not to look at just the config class? At a first glance, I'd say we want to compare the defaults to the class we instantiated, not to the superclass `PretrainedConfig`.

**Review by julien-c — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
> Any reason not to look at just the config class? At a first glance, I'd say we want to compare the defaults to the class we instantiated, not to the superclass `PretrainedConfig`.

Back then this was my initial idea as well - but then the configs could be more or less emtpy if all parameters are the same. This has a couple of disadvantages:
- When looking at the config online people cannot see any parameters and would have to look into the code which might be annoying
- This would make the configs much more prone to break if the init values of respective classes are changed.

**Comment by patrickvonplaten** [comment]
Update: This feature requires configs to be instantiated without providing any parameters. This is currently not possible for the `EncoderDecoderModelConfig` and `RagConfig` because those configs are composed of multiple sub-configs which have to be provided. => A new class attribute `is_composition` is added to correctly handle these classes.

File: src/transformers/configuration_fsmt.py [hunk 1]
```diff
@@ -126,9 +126,9 @@ class FSMTConfig(PretrainedConfig):
     # update the defaults from config file
     def __init__(
         self,
-        langs,
-        src_vocab_size,
-        tgt_vocab_size,
+        langs=["en", "de"],
```
Comment (patrickvonplaten) [inline src/transformers/configuration_fsmt.py:129]: IMO, it should be possible to instantiate every config (which is not a "composition" config) without providing any parameters: 

```python 
config = self.config_cls()
```

I added these init params from: https://huggingface.co/facebook/wmt19-en-de (cc @stas00)

File: tests/test_configuration_common.py [hunk 2]
```diff
@@ -66,9 +66,16 @@ def create_and_test_config_with_num_labels(self):
         self.parent.assertEqual(len(config.id2label), 3)
         self.parent.assertEqual(len(config.label2id), 3)
 
+    def check_config_can_be_init_without_params(self):
+        if self.config_class.is_composition:
+            return
+        config = self.config_class()
```
Comment (patrickvonplaten) [inline tests/test_configuration_common.py:72]: This makes sure that every config can be instantiated without providing any parameters.

**Comment by patrickvonplaten** [comment]
UPDATE: I had to add a class attribute to the config to make this feature work - @julien-c @sgugger @thomwolf @LysandreJik - could you check if this is fine for you guys.

File: src/transformers/configuration_utils.py [hunk 3]
```diff
@@ -41,6 +41,9 @@ class PretrainedConfig(object):
     Class attributes (overridden by derived classes)
         - **model_type** (:obj:`str`): An identifier for the model type, serialized into the JSON file, and used to
           recreate the correct object in :class:`~transformers.AutoConfig`.
+        - **is_composition** (:obj:`bool`): Whether the config class is composed of multiple
+          sub-configs. In this case the config has to be initialized from two or more configs of
+          type :class:`~transformers.PretrainedConfig`, *e.g.*: `EncoderDecoderConfig`, `RagConfig`.
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:46]:
```suggestion
          type :class:`~transformers.PretrainedConfig` like: :class:`~transformers.EncoderDecoderConfig` or
          :class:`~RagConfig`.
```

**Review by sgugger — APPROVED** [review]
Understood, this LGTM then! Thanks!

**Comment by julien-c** [comment]
LGTM

File: src/transformers/configuration_fsmt.py (hunk 1 above)
Comment (stas00) [reply src/transformers/configuration_fsmt.py:129]: These particular defaults make no sense w/o the corresponding vocab. Just as well set them to 1000 if defaults are now required. I remember some common tests fails if vocab size is less than 1000, so that is probably a good default.

File: src/transformers/configuration_fsmt.py [hunk 4]
```diff
@@ -126,9 +126,9 @@ class FSMTConfig(PretrainedConfig):
...
+        src_vocab_size=42024,
+        tgt_vocab_size=42024,
```
Comment (stas00) [inline src/transformers/configuration_fsmt.py:131]: The suggested vocab size defaults make no sense w/o the corresponding vocab. Just as well set them to 1000 if defaults are now required. I remember some common tests fails if vocab size is less than 1000, so that is probably a good default.

File: src/transformers/configuration_fsmt.py (hunk 1 above)
Comment (stas00) [reply src/transformers/configuration_fsmt.py:129]: I'd say, in such case let's use `["xx", "yy"]` so it's clear it's non-sense. Using meaningful data here is misleading at best.

File: src/transformers/configuration_fsmt.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/configuration_fsmt.py:131]: ok great! Thanks for the tip :-)

Comment (patrickvonplaten) [reply src/transformers/configuration_fsmt.py:131]: Sorry, actually I don't really see why 42024 makes no sense - could you explain a bit?

Comment (stas00) [reply src/transformers/configuration_fsmt.py:131]: Is this a set of defaults for the sake of default, or are we configuring a very specific tiny model as a default that is actually a working model?

Comment (stas00) [reply src/transformers/configuration_fsmt.py:131]: If you're using some real model as a default, then yes, you want the actual numbers. If these are just random numbers, then why would you want to set it to 42024?

Comment (stas00) [reply src/transformers/configuration_fsmt.py:131]: The "makes no sense" part is about a user reading model's default and scratching their head about why this particular number, or why these languages? Normally default values make certain sense. That's why I removed the defaults, since any defaults make no sense here.

The model can't do anything sensible with `en/de` without the specific model trained on this pair to support it. It can't do anything with the vocab_size of 42024 w/o that vocab. I hope this helps explaining it.

File: src/transformers/configuration_fsmt.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/configuration_fsmt.py:129]: But it's the actual `configuration_fsmt.py` file no? All other configuration files use the actual params of one of the main models as default. E.g. `BertConfig()` uses the params of `bert-base-cased` as defaults. I don't really see why this is misleading... @sgugger @LysandreJik what is your opinion here?

File: src/transformers/configuration_fsmt.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/configuration_fsmt.py:131]: Sorry answered above - yeah I'm in favor of actual numbers -> I don't think they should be random numbers.

Comment (stas00) [reply src/transformers/configuration_fsmt.py:131]: Thank you for explaining your point of view, @patrickvonplaten 

Because the configuration process for fsmt was complex, using real defaults originally masked problems, so by not having defaults it was detecting problems of not getting the right config immediately. I'm concerned that adding defaults that look about right all kinds of unexpected problems may arise.

My reasoning is this - if you don't have the actual model to support the defaults in config, then the defaults shouldn't be like in the model - I'm talking about the parts that are very specific to a model - like languages, or vocab  sizes. 

I'm not sure I'm being successful at explaining myself.

But regardless, if you feel that the defaults you suggested make more sense - please go for it. I trust you have more insight than I do here.

Comment (sshleifer) [reply src/transformers/configuration_fsmt.py:131]: I completely agree with @stas00 even though I have gotten yelled at for the same thing ;)
Makes no sense to have defaults that look like they might actually contain useful information when they don't. Just a distraction. I have the same view on documenting config defaults in docstrings.

Comment (stas00) [reply src/transformers/configuration_fsmt.py:131]: Another note:  I realized that you picked the default numbers from a tiny model which wasn't truly tiny ;) I'm working on a new version that is 50 times smaller, and there the vocab size is 21, instead of 42K. PR coming.

Comment (patrickvonplaten) [reply src/transformers/configuration_fsmt.py:131]: Hmm, interesting point of view! I guess for me it's quite natural that if you do:

```python 
config = BertConfig()
```

you would expect to get the most commonly used / standard BERT config being `bert-base-cased`, which is useful information IMO. The only other option that would make sense is to force the user to configure everything by himself, but I think putting random parameters there makes no sense. 

And since it's a big question of consistency with the library (we will not go back from having default configs I think), I see only the "real" default parameters option as the way to go. 

Would be keen to hear @sgugger and @LysandreJik opinion on that as well!

Comment (LysandreJik) [reply src/transformers/configuration_fsmt.py:131]: I think @patrickvonplaten has the right intentions  following what is established across the repo. In every configuration file (e.g., [BERT](https://github.com/huggingface/transformers/blob/master/src/transformers/configuration_bert.py#L55), [ALBERT](https://github.com/huggingface/transformers/blob/master/src/transformers/configuration_albert.py#L37), [CTRL](https://github.com/huggingface/transformers/blob/master/src/transformers/configuration_ctrl.py#L30-L31)), we have defaults for configurations so that initializing one without arguments yields a sensible architecture based on existing pre-trained weights. If it is not the case for a model, then it slipped the review and should be updated.

This means that doing: `BertModel(BertConfig())` yields a model similar to what the original BERT model is. This makes it easier to work with as we don't have a myriad of (usually different) arguments to supply to each configuration when doing tests.

Also, this is the convention we have adopted until now and I see no strong argument against it that would lead to changing this convention. I also see no argument on why FSMT would need to differ in that regard.

@stas00:
> Because the configuration process for fsmt was complex, using some defaults originally masked problems, so by not having defaults it was detecting problems of not getting the right config immediately. I'm concerned that adding defaults that look about right all kinds of unexpected problems may arise.

I would say that if some problems were masked by using some defaults, then some additional tests should be added to ensure that these tests are not an issue, either for this model or for others.

**Review by LysandreJik — APPROVED** [review]
LGTM!

File: src/transformers/configuration_fsmt.py (hunk 4 above)
Comment (stas00) [reply src/transformers/configuration_fsmt.py:131]: > this is the convention we have adopted until now and I see no strong argument against it that would lead to changing this convention

If you're saying that convention practically works, then it is.

> some additional tests should be added 

Yes, I have just done that here https://github.com/huggingface/transformers/pull/7860

Comment (sshleifer) [reply src/transformers/configuration_fsmt.py:131]: @LysandreJik The convention doesn't make as much sense for situations where you have many checkpoints trained from scratch, like FSMT/Marian. They all have different vocab sizes and none is meaningfully the "base" model in a `bert-base-cased` way.

Comment (sshleifer) [reply src/transformers/configuration_fsmt.py:131]: Similarly, there are two completely different blenderbot checkpoints and I arbitrarily chose the smaller to be the config defaults.

Comment (stas00) [reply src/transformers/configuration_fsmt.py:131]: As I mentioned twice earlier this is the way of the project, and so it is.

So I will just address @patrickvonplaten comment since i think I know why we are missing each other here:

>  I guess for me it's quite natural that if you do:
> 
> config = BertConfig()
> 
> you would expect to get the most commonly used / standard BERT config being bert-base-cased, which is useful information IMO.

I totally agree! It works for Bert and other similar models. It doesn't work for translation models, IMHO. 

What would as a user do with the default German to English translation model with 40K vocab of non-existing vocab -  the model the user will end up with will not be functional - yes, they have to configure it if they don't use a pretrained model. There is no magical way around it.

The key here is that certain models do not have sensible defaults and when this is the case giving a default that looks very much like correct default just for the consistency sake is questionable engineering-wise.

It'd work for automatic tests as in ThomWolf's recent PR as long as you don't try to do any qualitative tests, but it won't do anything useful for end users.
